README.md

```diff
@@ -0,0 +1,167 @@
+```
+
+### Custom types
+
+sanctuary-def does not use `instanceof` checks, as they are unacceptably
+fragile (see [plaid/sanctuary#100][5]). It's thus necessary to provide some
+mechanism other than constructor identity to determine whether a value is a
+member of a custom type. This is achieved through the use of a `'@@type'`
+property, which should use the form `'<package-name>/<type-name>'`, where
+`<package-name>` is the name of the npm package in which the type is defined.
+For example:
+
+```javascript
+function Gizmo(name) { this.name = name; }
+Gizmo.prototype['@@type'] = 'gadgets/Gizmo';
+
+//    gizmoName :: Gizmo -> String
+const gizmoName = def('gizmoName', [Gizmo], gizmo => gizmo.name);
+
+gizmoName(new Gizmo('thingamajig'));  // => 'thingamajig'
+```
+
+`gizmoName` requires an argument of type `Gizmo`. `new Gizmo('thingamajig')`
+is determined to be a member of this type due to this equivalence:
+
+```javascript
+new Gizmo('thingamajig')['@@type'] === Gizmo.prototype['@@type']
+```
+
+### Custom pseudotypes
+
+A type is essentially a set of values. Sometimes one wishes to define a type
+without creating a corresponding constructor function. These are referred to
+as pseudotypes in the Sanctuary world. JavaScript does not provide an Integer
+type, for example, but one could define an Integer pseudotype by defining the
+subset of Number values which are integers within the precisely representable
+range:
+
+```javascript
+//    Integer :: { name :: String, test :: a -> Boolean }
+const Integer = {
+  name: 'Integer',
+  test: x => Object.prototype.toString.call(x) === '[object Number]' &&
+             Math.floor(x) === Number(x) &&
+             x >= Number.MIN_SAFE_INTEGER &&
+             x <= Number.MAX_SAFE_INTEGER,
+};
+
+const even = def('even', [Integer], x => x % 2 === 0);
+
+even(100);  // => true
+even(123);  // => false
+even(0.5);  // ! TypeError: ‘even’ requires a value of type Integer as its first argument; received 0.5
+```
+
+A pseudotype definition is of type `{ name :: String, test :: a -> Boolean }`.
+
+### Type variables
+
+Polymorphism is powerful. It would be very limiting indeed to not be able to
+define a function for all types: one couldn't even define `identity`!
+
+One might want to define a `concat` function which concatenates two values of
+any one type which provides a `concat` method. `concat([1, 2, 3], [4, 5, 6])`
+should evaluate to `concat([1, 2, 3, 4, 5, 6])`; `concat('abc', 'def')` should
+evaluate to `'abcdef'`. It's nonsensical, though, to concatenate values of
+different types, so both arguments must be of the same type:
+
+```javascript
+const a = def.types.a;
+
+//    concat :: Semigroup a => a -> a -> a
+const concat = def('concat', [a, a], (x, y) => x.concat(y));
+```
+
+`def` has no notion of type classes. The `concat` function defined above does
+not ensure that its arguments are of a type which provides a `concat` method.
```

Comment:
Great point, @boxofrox. I agree that we should explain how we determine whether two values are of the same type.

Also, it might be clearer to give this paragraph its own subheading, __Type classes__, to avoid casting doubt on whether the `concat` definition above is valid use of `def`. So we would first introduce type variables, then ask the question of whether it's possible to constrain `a` by requiring it to implement [Semigroup][1], for example. The answer, currently, is no.


[1]: https://github.com/fantasyland/fantasy-land#semigroup